File: examples/load_boundaries.R
# Jordan Administrative Boundaries - R Usage Examples
#
# Demonstrates how to load, visualize, and analyze Jordan administrative boundaries
# using sf, ggplot2, and dplyr packages.
#
# Requirements:
#   install.packages(c("sf", "ggplot2", "dplyr", "readr"))

library(sf)
library(ggplot2)
library(dplyr)
library(readr)

# Set up paths (adjust to your local directory)
DATA_DIR <- "../data"
GPKG_DIR <- file.path(DATA_DIR, "gpkg")

# ==============================================================================
# Example 1: Load and display governorate boundaries
# ==============================================================================

cat("================================================================================\n")
cat("Example 1: Load Governorate Boundaries\n")
cat("================================================================================\n\n")

governorates <- st_read(file.path(GPKG_DIR, "gov_simpl_20m.gpkg"), quiet = TRUE)
cat(sprintf("Loaded %d governorates\n", nrow(governorates)))
cat(sprintf("CRS: %s\n\n", st_crs(governorates)$input))

# Display first few rows
cat("First 3 governorates:\n")
print(governorates %>% select(name_en, name_ar, wikidata) %>% head(3))

# Plot governorates
png("governorates_map_R.png", width = 10, height = 10, units = "in", res = 150)
plot(st_geometry(governorates), main = "Jordan Governorates",
     col = "lightblue", border = "black", lwd = 1.5)
# Add labels at centroids
centroids <- st_centroid(governorates)
text(st_coordinates(centroids), labels = governorates$name_en, cex = 0.7)
dev.off()
cat("Saved map to: governorates_map_R.png\n")

# ==============================================================================
# Example 2: Load and visualize districts
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 2: District Visualization with ggplot2\n")
cat("================================================================================\n\n")

districts <- st_read(file.path(GPKG_DIR, "dis_simpl_20m.gpkg"), quiet = TRUE)
cat(sprintf("Loaded %d districts\n", nrow(districts)))

# Create a simple map with ggplot2
p <- ggplot(districts) +
  geom_sf(aes(fill = name_en_2), color = "black", size = 0.3) +
  labs(title = "Jordan Districts by Governorate",
       fill = "Governorate") +
  theme_minimal() +
  theme(legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank())

ggsave("districts_by_governorate.png", p, width = 10, height = 12, dpi = 150)
cat("Saved map to: districts_by_governorate.png\n")

# ==============================================================================
# Example 3: Analyze district attributes
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 3: Analyze District Attributes\n")
cat("================================================================================\n\n")

# Count districts per governorate
districts_summary <- districts %>%
  st_drop_geometry() %>%
  group_by(name_en_2) %>%
  summarise(n_districts = n()) %>%
  arrange(desc(n_districts))

cat("Number of districts per governorate:\n")
print(districts_summary)

# Create a bar plot
p_bar <- ggplot(districts_summary, aes(x = reorder(name_en_2, n_districts), y = n_districts)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() +
  labs(title = "Number of Districts per Governorate",
       x = "Governorate",
       y = "Number of Districts") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10))

ggsave("districts_per_governorate_bar.png", p_bar, width = 8, height = 6, dpi = 150)
cat("\nSaved bar chart to: districts_per_governorate_bar.png\n")

# ==============================================================================
# Example 4: Spatial operations - Calculate area and buffer
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 4: Spatial Operations\n")
cat("================================================================================\n\n")

# Project to UTM Zone 36N for accurate area calculation
districts_utm <- st_transform(districts, crs = 32636)

# Calculate area in km²
districts_utm <- districts_utm %>%
  mutate(area_km2 = as.numeric(st_area(.)) / 1e6)

cat(sprintf("Total area of Jordan: %.0f km²\n", sum(districts_utm$area_km2)))
cat(sprintf("Mean district area: %.0f km²\n", mean(districts_utm$area_km2)))
cat(sprintf("Largest district: %.0f km²\n", max(districts_utm$area_km2)))
cat(sprintf("Smallest district: %.0f km²\n", min(districts_utm$area_km2)))

# Create 10km buffer around first district (for demonstration)
district_buffer <- st_buffer(districts_utm[1, ], dist = 10000)  # 10km in meters

# Plot original district and buffer
png("district_buffer_example.png", width = 8, height = 8, units = "in", res = 150)
plot(st_geometry(district_buffer), col = "lightblue", border = "blue",
     main = "10km Buffer Around Sample District")
plot(st_geometry(districts_utm[1, ]), col = "darkblue", border = "black", add = TRUE)
legend("topright", legend = c("Original District", "10km Buffer"),
       fill = c("darkblue", "lightblue"))
dev.off()
cat("\nSaved buffer example to: district_buffer_example.png\n")

# ==============================================================================
# Example 5: Spatial joins
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 5: Spatial Join - Districts to Governorates\n")
cat("================================================================================\n\n")

# Join districts to governorates (find which governorate each district is in)
districts_with_gov <- st_join(
  districts %>% select(name_en, geometry),
  governorates %>% select(gov_name = name_en, geometry),
  join = st_within
)

# Count districts per governorate
districts_per_gov <- districts_with_gov %>%
  st_drop_geometry() %>%
  count(gov_name, name = "n_districts") %>%
  arrange(desc(n_districts))

cat("Districts per governorate:\n")
print(districts_per_gov)

# ==============================================================================
# Example 6: Export to different formats
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 6: Export to Different Formats\n")
cat("================================================================================\n\n")

# Export to Shapefile
st_write(governorates, "governorates_R.shp", quiet = TRUE, delete_dsn = TRUE)
cat("Exported governorates to: governorates_R.shp\n")

# Export to GeoJSON
st_write(districts, "districts_R.geojson", driver = "GeoJSON", quiet = TRUE, delete_dsn = TRUE)
cat("Exported districts to: districts_R.geojson\n")

# Export to CSV (without geometry)
districts_no_geom <- districts %>%
  st_drop_geometry() %>%
  select(name_en, name_en_2, wikidata)

write_csv(districts_no_geom, "districts_attributes_R.csv")
cat("Exported district attributes to: districts_attributes_R.csv\n")

# ==============================================================================
# Example 7: Filter and subset
# ==============================================================================

cat("\n================================================================================\n")
cat("Example 7: Filter and Subset\n")
cat("================================================================================\n\n")

# Filter districts in Zarqa governorate
zarqa_districts <- districts %>%
  filter(name_en_2 == "Zarqa")

cat(sprintf("Found %d districts in Zarqa Governorate:\n", nrow(zarqa_districts)))
cat(paste("  -", zarqa_districts$name_en, collapse = "\n"), "\n")

# Plot Zarqa districts
p_zarqa <- ggplot(zarqa_districts) +
  geom_sf(aes(fill = name_en), color = "black", size = 0.5) +
  labs(title = "Districts in Zarqa Governorate") +
  theme_minimal() +
  theme(legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank())

ggsave("zarqa_districts.png", p_zarqa, width = 8, height = 8, dpi = 150)
cat("\nSaved Zarqa districts map to: zarqa_districts.png\n")

cat("\n================================================================================\n")
cat("All examples completed successfully!\n")
cat("================================================================================\n")

File: examples/load_boundaries.py
"""
Jordan Administrative Boundaries - Python Usage Examples

Demonstrates how to load, visualize, and analyze Jordan administrative boundaries
using GeoPandas, Matplotlib, and Pandas.

Requirements:
    pip install geopandas matplotlib pandas
"""

import geopandas as gpd
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path

# Set up paths (adjust to your local directory)
DATA_DIR = Path("../data")
GPKG_DIR = DATA_DIR / "gpkg"

# ==============================================================================
# Example 1: Load and display governorate boundaries
# ==============================================================================

print("=" * 80)
print("Example 1: Load Governorate Boundaries")
print("=" * 80)

governorates = gpd.read_file(GPKG_DIR / "gov_simpl_20m.gpkg")
print(f"\nLoaded {len(governorates)} governorates")
print(f"CRS: {governorates.crs}")
print(f"\nFirst 3 governorates:")
print(governorates[['name_en', 'name_ar', 'wikidata']].head(3))

# Plot governorates
fig, ax = plt.subplots(figsize=(10, 10))
governorates.plot(ax=ax, edgecolor='black', facecolor='lightblue', linewidth=1.5)
governorates.apply(lambda x: ax.annotate(text=x['name_en'],
                                          xy=x.geometry.centroid.coords[0],
                                          ha='center', fontsize=8), axis=1)
ax.set_title("Jordan Governorates", fontsize=14, fontweight='bold')
ax.axis('off')
plt.tight_layout()
plt.savefig("governorates_map.png", dpi=150)
print("\nSaved map to: governorates_map.png")

# ==============================================================================
# Example 2: Load and visualize districts
# ==============================================================================

print("\n" + "=" * 80)
print("Example 2: District Visualization")
print("=" * 80)

# Load districts (spatial)
districts = gpd.read_file(GPKG_DIR / "dis_simpl_20m.gpkg")
print(f"\nLoaded {len(districts)} districts from GPKG")
print(f"\nAvailable columns: {', '.join(districts.columns[:10])}...")

# Plot districts colored by governorate
fig, ax = plt.subplots(figsize=(12, 10))
districts.plot(
    column='name_en_2',  # Parent governorate name
    ax=ax,
    legend=True,
    cmap='tab12',
    edgecolor='black',
    linewidth=0.5,
    legend_kwds={'title': "Governorate", 'bbox_to_anchor': (1.05, 1), 'loc': 'upper left'}
)
ax.set_title("Jordan Districts Colored by Governorate",
             fontsize=14, fontweight='bold')
ax.axis('off')
plt.tight_layout()
plt.savefig("districts_by_governorate.png", dpi=150, bbox_inches='tight')
print("\nSaved district map to: districts_by_governorate.png")

# Print summary by governorate
print(f"\nDistricts per Governorate:")
districts_per_gov = districts.groupby('name_en_2').size().sort_values(ascending=False)
for gov, count in districts_per_gov.items():
    print(f"  {gov}: {count} districts")

# ==============================================================================
# Example 3: Spatial joins - Find which governorate each district belongs to
# ==============================================================================

print("\n" + "=" * 80)
print("Example 3: Spatial Join (Districts to Governorates)")
print("=" * 80)

# Perform spatial join
districts_with_gov = gpd.sjoin(
    districts[['name_en', 'geometry']],
    governorates[['name_en', 'geometry']],
    how='left',
    predicate='within'
)

# Show first few results
print("\nDistricts and their parent governorates:")
print(districts_with_gov[['name_en_left', 'name_en_right']].head(10).to_string(index=False))

# ==============================================================================
# Example 4: Calculate area and population density
# ==============================================================================

print("\n" + "=" * 80)
print("Example 4: Calculate Area and Population Density")
print("=" * 80)

# Project to UTM Zone 36N (EPSG:32636) for accurate area calculation
districts_utm = districts.to_crs("EPSG:32636")

# Calculate area in km²
districts_utm['area_km2'] = districts_utm.geometry.area / 1_000_000

# Merge with population data
if 'districts_for_suave_with100K_2_no_wkt_Population 2024#number' in districts.columns:
    districts_utm['population'] = districts['districts_for_suave_with100K_2_no_wkt_Population 2024#number']
    districts_utm['pop_density_per_km2'] = districts_utm['population'] / districts_utm['area_km2']

    print(f"\nArea and Population Density Statistics:")
    print(f"  Total area: {districts_utm['area_km2'].sum():.0f} km²")
    print(f"  Mean district area: {districts_utm['area_km2'].mean():.0f} km²")
    print(f"  Total population: {districts_utm['population'].sum():,.0f}")
    print(f"  Mean population density: {districts_utm['pop_density_per_km2'].mean():.0f} per km²")

    # Find most and least dense districts
    most_dense = districts_utm.nlargest(3, 'pop_density_per_km2')
    print(f"\nMost densely populated districts:")
    for idx, row in most_dense.iterrows():
        print(f"  {row['name_en']}: {row['pop_density_per_km2']:.0f} per km²")

# ==============================================================================
# Example 5: Export to other formats
# ==============================================================================

print("\n" + "=" * 80)
print("Example 5: Export to Other Formats")
print("=" * 80)

# Export governorates to Shapefile
governorates.to_file("governorates.shp")
print("Exported governorates to: governorates.shp")

# Export districts to GeoJSON
districts.to_file("districts.geojson", driver="GeoJSON")
print("Exported districts to: districts.geojson")

# Export district attributes to CSV (without geometry)
districts_df = pd.DataFrame(districts.drop(columns='geometry'))
districts_df.to_csv("districts_attributes.csv", index=False)
print("Exported district attributes to: districts_attributes.csv")

# ==============================================================================
# Example 6: Filter and subset
# ==============================================================================

print("\n" + "=" * 80)
print("Example 6: Filter Districts by Governorate")
print("=" * 80)

# Filter districts in Amman governorate
amman_districts = districts[districts['name_en_2'] == 'Amman']
print(f"\nFound {len(amman_districts)} districts in Amman Governorate:")
print(amman_districts['name_en'].tolist())

# Plot only Amman governorate districts
fig, ax = plt.subplots(figsize=(10, 10))
amman_districts.plot(ax=ax, edgecolor='black', cmap='Set3')
ax.set_title("Districts in Amman Governorate", fontsize=14, fontweight='bold')
ax.axis('off')
plt.tight_layout()
plt.savefig("amman_districts.png", dpi=150)
print("\nSaved Amman districts map to: amman_districts.png")

print("\n" + "=" * 80)
print("All examples completed successfully!")
print("=" * 80)
